src/ui/veles_mainwindow.cc

```diff
@@ -413,6 +590,66 @@ DockWidget* MainWindowWithDetachableDockWidgets::findDockNotTabifiedWith(
   return nullptr;
 }
 
+void MainWindowWithDetachableDockWidgets::setDockWidgetsWithNoTitleBars(
+    bool no_title_bars) {
+  dock_widgets_with_no_title_bars_ = no_title_bars;
+  updateDockWidgetTitleBars();
+}
+
+bool MainWindowWithDetachableDockWidgets::dockWidgetsWithNoTitleBars() {
+  return dock_widgets_with_no_title_bars_;
+}
+
+QDockWidget* MainWindowWithDetachableDockWidgets::tabToDockWidget(
+    QTabBar* tab_bar, int index) {
+  if(tab_bar) {
+    // Based on undocumented feature (tested with Qt 5.5/5.7).
+    // QTabBars that control visibility of QDockWidgets grouped in
+    // QDockWidgetGroupWindows hold ids of respective QDockAreaLayoutInfos
+    // in tabData. Conveniently those ids are just pointers to
+    // QDockWidgets (stored as quintptr) and can be retrieved through public
+    // interface.
+    QDockWidget* dock =
+        reinterpret_cast<QDockWidget*>(
+        qvariant_cast<quintptr>(tab_bar->tabData(index)));
+
+    return dock;
+  }
+
+  return 0;
+}
+
+QTabBar* MainWindowWithDetachableDockWidgets::dockWidgetToTab(
+    QDockWidget* dock_widget) {
+  for (auto tab_bar : findChildren<QTabBar*>()) {
+    for (int i = 0; i < tab_bar->count(); ++i) {
+      if (tabToDockWidget(tab_bar, i) == dock_widget) {
+        return tab_bar;
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+MainWindowWithDetachableDockWidgets* MainWindowWithDetachableDockWidgets
+    ::getParentMainWindow(QObject* obj) {
+  while(true) {
+    if(obj == nullptr) {
+      return nullptr;
+    }
+
+    auto main_window = dynamic_cast<MainWindowWithDetachableDockWidgets*>(obj);
+    if(main_window) {
+      return main_window;
+    }
+
+    obj = obj->parent();
+  }
+
+  return nullptr;
```

Comment:
I do agree. This additional return statement was meant to avoid compiler warnings.